File: Grind75/C++/validParenthesis.cpp
class Solution {
public:
    bool isValid(string s) {
        if (s.length() == 0)
            return true;
        stack<char>stk;
        for (char ch : s) {
            if(ch == '(' or ch == '[' or ch == '{')     
                stk.push(ch);
            else if(ch == ')' and !stk.empty() and stk.top() == '(')
                stk.pop();
            else if(ch == '}' and !stk.empty() and stk.top() == '{')
                stk.pop();
            else if(ch == ']' and !stk.empty() and stk.top() == '[')
                stk.pop();
            else
                return false;
        }
        return stk.empty() ? true : false;
    }
};

File: valid_parenthesis.py
class Solution:
    def isValid(self, s: str) -> bool:
        stack = deque()
        for char in s:
            if char == '{' or char == '[' or char == '(':
                stack.append(char)
            elif len(stack) > 0 and char == '}' and stack[-1] == '{':
                stack.pop()
            elif len(stack) > 0 and char == ')' and stack[-1] == '(':
                stack.pop()
            elif len(stack) > 0 and char == ']' and stack[-1] == '[':
                stack.pop()
            else:
                return False
        
        return len(stack) == 0
            